pkg/codegen/dotnet/gen_program_expressions.go

```diff
@@ -0,0 +1,560 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dotnet
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/schema"
+	"io"
+	"math/big"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2"
+	"github.com/pulumi/pulumi/pkg/v2/codegen/hcl2/model"
+	"github.com/pulumi/pulumi/sdk/v2/go/common/util/contract"
+	"github.com/zclconf/go-cty/cty"
+)
+
+type nameInfo int
+
+func (nameInfo) Format(name string) string {
+	return validIdentifier(name)
+}
+
+// rewriteExpression amends the expression with intrinsics for C# generation.
+func (g *generator) rewriteExpression(expr model.Expression, typ model.Type) model.Expression {
```

Comment:
Let's call this `lowerExpression` : the rewrites we're doing are consistent with the general definition of lowering as a process of rewrites that prepare a program for code generation, and `lowerExpression` is used in the other backends.

```suggestion
func (g *generator) lowerExpression(expr model.Expression, typ model.Type) model.Expression {
```

Reply:
Sure, that reminds me to ask you to talk me through what "lower" means here (hopefully I'll be able to join the call later today)